lib/Frontend/DependencyVerifier.cpp

```diff
@@ -0,0 +1,680 @@
+  MATRIX_ENTRY("expected-private-member", Private, Member)                     \
+  MATRIX_ENTRY("expected-cascading-member", Cascading, Member)                 \
+  MATRIX_ENTRY("expected-private-dynamic-member", Private, DynamicMember)      \
+  MATRIX_ENTRY("expected-cascading-dynamic-member", Cascading, DynamicMember)
+struct Expectation {
+public:
+  enum class Kind : uint8_t {
+    Negative,
+    Provides,
+    Member,
+    PotentialMember,
+    Superclass = PotentialMember,
+    Conformance = PotentialMember,
+    DynamicMember,
+  };
+
+  enum class Flavor : uint8_t {
+    None,
+    Private,
+    Cascading,
+  };
+
+  /// The full range of the "expected-foo {{}}".
+  const char *ExpectedStart, *ExpectedEnd = nullptr;
+
+  /// Additional information about the expectation.
+  std::pair<Expectation::Kind, Expectation::Flavor> Info;
+
+  /// The raw input buffer for the message text, the part in the {{...}}
+  StringRef MessageRange;
+
+public:
+  Expectation(const char *ExpectedStart, Expectation::Kind k,
+              Expectation::Flavor f)
+      : ExpectedStart(ExpectedStart), Info{k, f} {}
+
+  bool isCascading() const {
+    return Info.second == Expectation::Flavor::Cascading;
+  }
+};
+
+/// An \c Obligation represents a compiler-provided entry in the set of
+/// dependencies for a given source file. Similar to an \c Expectation an \c
+/// Obligation contains a name, information about its kind and flavor, and an
+/// extra source location that can be used to guide where diagnostics are
+/// emitted. Unlike an \c Obligation, it provides an extra piece of state that
+/// represents the obligation's "fulfillment status".
+///
+/// All \c Obligations begin in the \c Active state. Once an obligation has been
+/// paired with a matching \c Expectation, the obligation may then transition to
+/// either \c Fulfilled if it has been satisfied completely, or \c Failed
+/// otherwise. The verifier turns all unfulfilled obligations into errors.
+struct Obligation {
+  /// The state of an \c Obligation
+  enum class State : uint8_t {
+    /// The \c Obligation is active and has not been paired with a corresponding
+    /// \c Expectation.
+    Active,
+    /// The \c Obligation is fulfilled.
+    Fulfilled,
+    /// The \c Obligation was matched against an \c Expectation, but that
+    /// expectation could not
+    /// fulfill the obligation because additional requirements did not pass.
+    Failed,
+  };
+
+  /// A token returned when an \c Obligation is fulfilled or failed. An \c
+  /// Obligation is the only type that may construct fulfillment tokens.
+  struct FullfillmentToken {
+    friend Obligation;
+
+  private:
+    FullfillmentToken() = default;
+  };
+
+  /// An \c Obligation::Key is a reduced set of the common data contained in an
+  /// \c Obligation and an \c Expectation.
+  ///
+  /// This provides a way to use a value of either type to index into an  \c
+  /// ObligationMap.
+  struct Key {
```

Comment:
Negative expectations are kind of special under the current scheme since they're currently a global assertion ("there is no dependency of any kind or scope with this key". Hence, the `NegativeExpectationMap` doesn't need this abstraction for now.

Reply:
OK